stringdtype/stringdtype/src/umath.c

```diff
@@ -13,6 +13,95 @@
 #include "string.h"
 #include "umath.h"
 
+static int
+minmax_resolve_descriptors(struct PyArrayMethodObject_tag *NPY_UNUSED(method),
+                           PyArray_DTypeMeta *NPY_UNUSED(dtypes[]),
+                           PyArray_Descr *given_descrs[],
+                           PyArray_Descr *loop_descrs[],
+                           npy_intp *NPY_UNUSED(view_offset))
+{
+    Py_INCREF(given_descrs[0]);
+    loop_descrs[0] = given_descrs[0];
+    Py_INCREF(given_descrs[1]);
+    loop_descrs[1] = given_descrs[1];
+
+    StringDTypeObject *new = new_stringdtype_instance();
+    if (new == NULL) {
+        return (NPY_CASTING)-1;
+    }
+    loop_descrs[2] = (PyArray_Descr *)new;
+
+    return NPY_NO_CASTING;
+}
+
+// The strided loop is called from ufunc_object.c.
+// If this is called from `reduce_loop()`, data[0] and data[2] are the same
+// pointer, and it is pointed at the single element in the output array - the
+// accumulator; the stride will be zero in this case.
+static int
+maximum_strided_loop(PyArrayMethod_Context *NPY_UNUSED(context),
+                     char *const data[], npy_intp const dimensions[],
+                     npy_intp const strides[], NpyAuxData *NPY_UNUSED(auxdata))
+{
+    npy_intp N = dimensions[0];
+    char *in1 = data[0];
+    char *in2 = data[1];
+    char *out = data[2];
+    npy_intp in1_stride = strides[0];
+    npy_intp in2_stride = strides[1];
+    npy_intp out_stride = strides[2];
+
+    while (N--) {
+        if (compare(in1, in2, NULL) > 0) {
+            if (ssset((ss *)in1, (ss *)out) < 0) {
+                return -1;
+            }
+        }
+        else {
+            if (ssset((ss *)in2, (ss *)out) < 0) {
+                return -1;
+            }
+        }
```

Comment:
I wonder if it would be more performant for reductions to first scan over the input array to find the largest possible string, allocate a temporary string to hold the output, and then instead of doing `malloc/free` over and over for the output just re-use the temporary string buffer. Then when the loop is over, allocate a new `ss` string to hold the result and set it to the output buffer.

I worry that this implementation will lead to memory thrashing in the worst case.

Reply:
Agreed, in fact that's what we do for the `argmin`/`argmax` array funcs. In this case, `maximum_strided_loop` is called for two different purposes which are somewhat in tension here:

1. Inside a reduce function in the case where `arr.max()` is desired
2. When `np.maximum(x, y)` is called, where the return type is an array

In the second case, we want the current behavior. In the first case, we can just loop over the array and find the max without doing `malloc/free` for each element. One thing we could do is detect when we are doing `arr.max` by not doing `malloc/free` in cases where `in1_stride == 0`, since that only arises from the accumulator array in the reduce function.

Reply:
I would say the typical approach for this would be to:

1. `get_loop` can attach an `auxdata` with some minimal string size already allocated (can be 0 also).
2. Whenever you need more space, you can have an `unlikely` marked code path to grow that space (`unlikely` may be helpful if it isn't a non-inlined function call anyway).
   * A typical strategy would probably over-allocate a bit since it doesn't hurt (for example, I have something like this in NumPy as `grow_size_and_multiply`, but there are more)
3. auxdata free has to free the space obviously.

This is single pass.  A single length check should be fast (and with overallocation fail rarely).  We just want to avoid that we don't have a lot of unused garbage code in the inner-loop (thus the `unlikely`).

Reply:
Wooops, sorry... I suppose for a reduction like this, you want to work "in-place".

So not sure this works super nicely/straight forward.  I suppose this is part of the reason why many string implementations have not just a length but also a "capacity", which allows to overallocate and re-use existing space liberally.